### ISSUE TITLE
Blocks entire GoCD server when Github is down/slow

### ISSUE THREAD

**Issue by dziemba** [gocd-contrib/github-oauth-authorization-plugin#14]
When Github is down / unresponsive, it looks like the Github OAuth flow is blocking
the GoCD webserver, causing it to not respond to any more requests.
Looking at the threaddump, it seems like all webserver threads are hanging,
waiting for Github to respond.

The server also didn't respond to SIGTERM anymore. After killing it and bringing it back up, it
works as expected again.

Threaddump attached. Please let me know if you need more info.

## Versions

os: debian stretch
gocd: 18.5.0
github-oauth-authorization-plugin: 2.0.0-12
other plugins: github-pr-poller 1.3.4, github-pr-status: 1.4

[gocd-threaddump.txt.gz](https://github.com/gocd-contrib/github-oauth-authorization-plugin/files/2075789/gocd-threaddump.txt.gz)

http://fastthread.io/my-thread-report.jsp?p=c2hhcmVkLzIwMTgvMDYvNi8tLXgtLTEwLTI5LTUy

**Comment by ketan** [comment]
> The server also didn't respond to SIGTERM anymore. After killing it and bringing it back up, it works as expected again.

I hope it was not a `kill -9` that has the potential to corrupt your (h2) DB.

**Comment by dziemba** [comment]
> I hope it was not a kill -9 that has the potential to corrupt your (h2) DB.

Yes, I had to `kill -9` and was aware of the potential issues (in the worst case we have backups...). It didn't respond to TERM or anything else for minutes and we had to try to bring it back up. Seems to be fine now. Is there a better way to force a clean shutdown?
